### PR TITLE
Atomic gemm for TP-AR and TP-RS overlap with P2P exchanges

### PR DESCRIPTION
**Pull request by erhoo82**
- Atomic GEMM for TP-AG and TP-RS overlaps with P2P ring-exchanges
- TP-AG overlap shuffles GEMM chunk outputs and TP-RS overlap re-shuffles. So, both TP-AG and TP-RS should use `atomic GEMM` and `ring_exhcnage` for functionality.
- `NVTE_AG_P2P_MULTI_ATOMIC=1` uses a single kernel for TP-AG.
- Updated the assert conditions of userbuffer initialization.
- Update the default value of `num_splits` to 4 from 0, which is only valid for TP-RS overlap with `pipeline` overlap.

### CODE REVIEW

**Comment by erhoo82** [comment]
@timmoon10 
Can you trigger a ci test?

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -160,20 +166,35 @@ def add_ub(
...
-                "`pipeline` overlap method is not supported for AllGather."
+                f"At {name}, `pipeline` overlap method is not supported for AllGather."
+            )
+        # Check if both AG and RS overlaps use `atomic GEMM`` + `p2p ring-exchange`.
+        # Using atomic GEMM + p2p ring-exchange in only one of the pair breaks functionality.
+        global layers_atomic_ring_exchange
+        if atomic_gemm and method == "ring_exchange" and name in ag_rs_pairs:
+            layers_atomic_ring_exchange += [name, ag_rs_pairs[name]]
+        if name in rs_ag_pairs:
+            assert_massage = (
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/base.py:185]: This error is not very clear:
```suggestion
            assert_message = (
```

**Review by timmoon10 — APPROVED** [review]
LGTM once the tests pass.

**Comment by erhoo82** [comment]
Fixed lint errors and reflected comments.
@timmoon10  Please trigger ci.

**Comment by erhoo82** [comment]
@timmoon10 Can we merge this PR?

**Pull request merged**